### PR TITLE
symex renaming of mathematical functions

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
This is failing one test in CI.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3fd03709115c6d70d64641365a94c27a41b68582).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102946505

**Review by tautschnig — APPROVED** [review]

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 1]
```diff
@@ -267,24 +269,31 @@ static auxiliary_symbolt declare_function(
   return func_symbol;
 }
 
-/// Create a function application expression.
+/// Create a (mathematical) function application expression.
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp:272]: Maybe rephrase that, for instance to say "a function application expression with mathematical function type"

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -165,8 +165,12 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
 {
   const function_application_exprt &function_app =
     to_function_application_expr(expr);
-  irep_idt func_id =
-    function_app.function().find(ID_expression).get(ID_identifier);
+
+  if(function_app.function().id() != ID_symbol)
+    return true;
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp:170]: I'm not sure this is in the right commit

File: src/goto-symex/goto_symex_state.cpp [hunk 3]
```diff
@@ -333,9 +335,13 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
...
   else if(expr.id()==ID_symbol)
   {
+    const auto &type = as_const(expr).type();
+
     // we never rename function symbols
-    if(as_const(expr).type().id() == ID_code)
+    if(type.id() == ID_code || type.id() == ID_mathematical_function)
+    {
       rename<level>(expr.type(), to_symbol_expr(expr).get_identifier(), ns);
+    }
```
Comment (romainbrenguier) [inline src/goto-symex/goto_symex_state.cpp]: introducing these brackets is unnecessary

**Review by romainbrenguier — APPROVED** [review]
mathematical_function_typet could benefit from a bit more documentation, because the difference with other function types is not clear. But the changes in this PR look good.

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 1 above)
Comment (kroening) [reply jbmc/src/java_bytecode/java_utils.cpp:272]: The function application expression doesn't have the type of the function, but the type of the codomain.
I've tweaked the comment to say that.

File: src/util/simplify_expr.cpp (hunk 2 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:170]: It does seem to be the most relevant of the three; it could perhaps be a separate one.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: bd34427d6cecab1d017f75e6604814cb0d3f675b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103031035
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 4]
```diff
@@ -267,24 +269,32 @@ static auxiliary_symbolt declare_function(
...
+/// The application has the type of the codomain (range) of the function.
 /// \param function_name: the name of the function
-/// \param arguments: a list of arguments
-/// \param type: return type of the function
+/// \param arguments: a list of arguments (an element of the domain)
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_utils.cpp:275]: elements of the domain

File: src/util/simplify_expr.cpp [hunk 5]
```diff
@@ -165,8 +165,12 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+
+  const irep_idt func_id =
```
Comment (smowton) [inline src/util/simplify_expr.cpp:172]: const ref?

**Review by smowton — APPROVED** [review]

File: src/util/simplify_expr.cpp (hunk 5 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:172]: Deliberately didn't -- a problem in the design of the simplifier is that it takes a non-const ref to an expression, and then changes that expression (say as opposed to building a new expression and then returning it). This may well invalidate those const refs taken at the beginning. With this design, it's safer to copy.

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 4 above)
Comment (kroening) [reply jbmc/src/java_bytecode/java_utils.cpp:275]: No, the arguments as a whole are an element of the domain; each argument is an element of the projection of the domain to the member of the tuple type.
